influx-reports/influx.py

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python3
+import argparse
+import datetime
+import dateutil.parser
+import logging
+import os
+import sys
+import yaml
+from collections import defaultdict
+from influxdb import InfluxDBClient
+from subunit.v2 import StreamResultToBytes
+
+
+def return_time(client, query, delta_seconds=0):
+    """ From an InfluxDB query, fetch
+    the first point time, and return a
+    python time object. Shift it from
+    a few seconds (delta_seconds) if necessary.
+    """
+    timestamp_query = client.query(query)
+    # Get points is generator, we should just get first
+    # point time (string type).
+    fpt_str = next(timestamp_query.get_points())['time']
+    try:
+        fpt = dateutil.parser.parse(fpt_str) + datetime.timedelta(seconds=delta_seconds)
+    except Exception as e:
+        raise SystemExit(e)
+    return fpt
+
+
+def return_uptime(client, table, job_reference, resolution=60):
+    """ For a certain build (job_reference), estimate the amount
+    of seconds a component (column of table) was up.
+    For that, we slice the job by resolution.
+    If I have no data in the slice, I assume downtime. If I have
+    at least one data per slice, I assume up, even if degraded.
+    The count of these positive events multiplied by resolution
+    is an average idea of the uptime.
+    """
+    uptimes = defaultdict(int)
+
+    query = ("select max(/.*_status/) from {} where time < now() and "
```

Comment:
I believe @hughsaunders is referring to the replacement fields `{}`, they are currently unnamed. For example `"{} - {named}".format("I have no name", named="I have a name.")`.